src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs

```diff
@@ -41,7 +41,7 @@ public static class ResolverUtility
             var requiredPackageIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
             requiredPackageIds.UnionWith(newPackageIdSet);
             requiredPackageIds.UnionWith(installedPackageIds);
-            var requiredPackages = solution.Where(package => requiredPackageIds.Contains(package.Id)).ToList();
+            var requiredPackages = solution.Where(package => requiredPackageIds.Contains(package.Id));
```

Comment:
This would cause the where clause run repeatedly because the requiredPackages is enumerated in a loop.